src/Microsoft.Net.Sdk.AnalyzerRedirecting/SdkAnalyzerAssemblyRedirector.cs

```diff
@@ -0,0 +1,142 @@
+            {
+                foreach (string analyzerPath in Directory.EnumerateFiles(versionDirectory, "*.dll", SearchOption.AllDirectories))
+                {
+                    if (!analyzerPath.StartsWith(versionDirectory, StringComparison.OrdinalIgnoreCase))
+                    {
+                        continue;
+                    }
+
+                    string version = Path.GetFileName(versionDirectory);
+                    string analyzerName = Path.GetFileNameWithoutExtension(analyzerPath);
+                    string pathSuffix = analyzerPath.Substring(versionDirectory.Length + (EndsWithSlash(versionDirectory) ? 0 : 1));
+                    pathSuffix = Path.GetDirectoryName(pathSuffix);
+
+                    AnalyzerInfo analyzer = new() { FullPath = analyzerPath, ProductVersion = version, PathSuffix = pathSuffix };
+
+                    if (builder.TryGetValue(analyzerName, out var existing))
+                    {
+                        existing.Add(analyzer);
+                    }
+                    else
+                    {
+                        builder.Add(analyzerName, [analyzer]);
+                    }
+                }
+            }
+        }
+
+        return builder.ToImmutable();
+    }
+
+    public string? RedirectPath(string fullPath)
+    {
+        if (AnalyzerMap.TryGetValue(Path.GetFileNameWithoutExtension(fullPath), out var analyzers))
+        {
+            foreach (AnalyzerInfo analyzer in analyzers)
+            {
+                var directoryPath = Path.GetDirectoryName(fullPath);
+                if (endsWithIgnoringTrailingSlashes(directoryPath, analyzer.PathSuffix) &&
+                    majorAndMinorVersionsMatch(directoryPath, analyzer.PathSuffix, analyzer.ProductVersion))
+                {
+                    return analyzer.FullPath;
+                }
+            }
+        }
+
+        return null;
+
+        static bool majorAndMinorVersionsMatch(string directoryPath, string pathSuffix, string version)
+        {
+            // Find the version number in the directory path - it is in the directory name before the path suffix.
+            // Example:
+            // "C:\Program Files\dotnet\packs\Microsoft.WindowsDesktop.App.Ref\8.0.8\analyzers\dotnet\" = directoryPath
+            //                                                                       ~~~~~~~~~~~~~~~~   = pathSuffix
+            //                                                                 ~~~~~                    = directoryPathVersion
+            int index = directoryPath.LastIndexOf(pathSuffix, StringComparison.OrdinalIgnoreCase);
+            if (index < 0)
+            {
+                return false;
+            }
+            string directoryPathVersion = Path.GetFileName(Path.GetDirectoryName(directoryPath.Substring(0, index)));
+
+            return getMajorMinorPart(directoryPathVersion) == getMajorMinorPart(version);
```

Comment:
This is reasonable. I was imagining you'd use `System.Version` but I can see how that might complicate things.  If you leave it this way you can avoid creating new strings by having a single method that compares the strings.  Here's a [sample](https://sharplab.io/#v2:EYLgHgbALANALiATgVwHYwCYgNQB8CwAUAAIBMAjEUceQJwAUAxgPYC2ADgIaICmAspwBWzRHwCWqEQAVucegCJaAOgAMqgLTteANzE8A7koDsSgBwQTAVgDMK2vJgACRao1aeug0vJKIli/IAlIEA3NR0TGxcvALCohLSsgqmrioOzspqKpo6eoY+fgHBYSQRLBzc/EIi4pKIMohyLmrk6c3eacVUhOEQjsDMzAA2juXRVXG1iY30NCqO2jyIAM5izKjkTnMLS6vrpIFEAN5EjmeOEnCOAGZiK3AAIsxwAJKoGDxgjgC8OytrGyUbw+YAA8td6AByJSQ0Knc5ia6Oei3e5PV7vT6OAA8jhUh0I50cJ0JRPOxCMN04Q2WPBKRIAvt0iZdHLSWO90cCsb9Fv91j5uWCIdDIU5UctHs8hY5sI5yHDSWdEcj2esMFzMV9cfj4WcSWTyZTrtTafTzkyekrHBS8T9fnMlABhKKVeh8vYbJwqJwegGkb1ONWc6VapwAZTgiAkAHMXRVo8t1kpQYgMBJqS8Y3UeE7OLTFQygA==)

Reply:
Also `System.Version` cannot parse versions with a hyphen it seems.